google/cloud/spanner/timestamp.cc

```diff
@@ -52,347 +47,66 @@ Status NegativeOverflow(std::string const& type) {
-      return NegativeOverflow(kDestType);
-    }
-  }
-  sec *= numerator;
-
-  auto sscount = count - sec * denominator / numerator;
-  if (denominator > kNanosPerSecond) {
-    auto const divider = denominator / kNanosPerSecond;
-    if (sscount < 0) sscount -= divider - 1;  // floor
-    return FromCounts(sec, sscount / divider * numerator);
-  }
-  auto const multiplier = kNanosPerSecond / denominator;
-  return FromCounts(sec, sscount * multiplier * numerator);
-}
-
-// [sec, nsec] => bounded (count * numerator/denominator) seconds.
-//
-// Only designed to handle the ratios of the std::chrono::duration helper
-// aliases, where either the numerator or the denominator is 1, and where
-// subsecond ratios are powers of 10.
-StatusOr<std::intmax_t> Timestamp::ToRatio(
-    std::intmax_t const min, std::intmax_t const max,
-    std::intmax_t const numerator, std::intmax_t const denominator) const {
-  constexpr auto kDestType = "std::chrono::time_point";
-
-  auto count = sec_ / numerator;
-  if (count >= 0) {
-    if (count > std::numeric_limits<std::intmax_t>::max() / denominator) {
-      return PositiveOverflow(kDestType);
-    }
-  } else {
-    if (count < std::numeric_limits<std::intmax_t>::min() / denominator) {
-      // Might be premature to declare overflow on an intermediate value.
-      return NegativeOverflow(kDestType);
-    }
-  }
-  count *= denominator;
-
-  auto ncount = nsec_ / numerator;
-  if (denominator < kNanosPerSecond) {
-    ncount /= kNanosPerSecond / denominator;
-  } else {
-    ncount *= denominator / kNanosPerSecond;
-  }
-
-  if (count > std::numeric_limits<std::intmax_t>::max() - ncount) {
-    return PositiveOverflow(kDestType);
-  }
-  count += ncount;
-
-  if (count > max) return PositiveOverflow(kDestType);
-  if (count < min) return NegativeOverflow(kDestType);
-  return count;
+std::ostream& operator<<(std::ostream& os, Timestamp ts) {
+  return os << internal::TimestampToRFC3339(ts);
 }
 
 namespace internal {
 
+// Timestamp objects are always formatted in UTC, and we always format them
+// with a trailing 'Z'. However, we're a bit more liberal in the UTC offsets we
+// accept, thus the use of '%Ez' in kParseSpec.
+auto constexpr kFormatSpec = "%E4Y-%m-%dT%H:%M:%E*SZ";
+auto constexpr kParseSpec = "%Y-%m-%dT%H:%M:%E*S%Ez";
+
 StatusOr<Timestamp> TimestampFromRFC3339(std::string const& s) {
-  return Timestamp::FromRFC3339(s);
+  absl::Time t;
+  std::string err;
+  if (absl::ParseTime(kParseSpec, s, &t, &err)) return MakeTimestamp(t);
+  return InvalidArgument(s + ": " + err);
 }
 
-std::string TimestampToRFC3339(Timestamp ts) { return ts.ToRFC3339(); }
+std::string TimestampToRFC3339(Timestamp ts) {
+  auto const t = ts.get<absl::Time>().value();  // Cannot fail.
+  return absl::FormatTime(kFormatSpec, t, absl::UTCTimeZone());
+}
 
+// Note: This function requires the `proto` argument to be within the
+// documented valid range for `protobuf::Timestamp`.
```

Comment:
So, if we get something out of range off the wire we crash?

Reply:
I was contemplating how best to handle this. Last night I was thinking about changing `ToAbslTime()` to automatically clamp the input proto values to the bound. However, this morning that seemed like an undesirable solution to me because then we'd be silently accepting and operating on invalid protos.

The option I liked better this morning (and I changed the code to do this) is that I changed this function (`TimestampFromProto`) to return `StatusOr<Timestamp>`, and I updated a handful of callers to check the returned status. This change makes sense to me because it matches our other functions that construct `Timestamp`, all of which might fail and return `StatusOr<Timestamp>`. Also, we're propagating the error to the caller so they can decide how to handle it if they want. That seems better than crashing, or silently using an invalid proto.